chore: update README, explaining new features (#347)
cdklabs/decdk: README.md

```

As you can see, the deCDK has the same semantics as a CloudFormation template. It contains a section for "Resources",
where each resource is defined by a *type* and a set of *properties*. deCDK allows using constructs from AWS Construct
Library in templates by identifying the class name (in this case `aws-cdk-lib.aws_sqs.Queue`).

When deCDK processes a template, it identifies these special resources and under-the-hood, it instantiates an object of
that type, passing in the properties to the object's constructor. All CDK constructs have a uniform signature, so this
is actually straightforward.

## Development

### Examples/Tests

When you build this module, it will produce a `cdk.schema.json` file at the root, which is referenced by the examples in
the [`examples`](./examples) directory. This directory includes working examples of deCDK templates for various areas.
We also snapshot-test those to ensure there are no unwanted regressions.

### Running local code

You can execute `decdk` using the current TypeScript code, without having to build the project. Provide this as your app
argument to `cdk`:

```sh
cdk -a "ts-node src/decdk.ts template.json" synth
```

> ## Other features

Reviewer: ```suggestion
## Features
```

> ### Raw CloudFormation

Reviewer: ```suggestion
### CloudFormation Resources
```

uniquely determine which construct you are trying to declare. But the return type of a method call is unambiguous and
can be inferred by deCDK. Nonetheless, if you still want to declare the type, as a form of
documentation, it's perfectly valid to do so. Just keep in mind that the declared type must match the method's
return type. Otherwise, deCDK will stop synthesis and give you an error message.

The second point to notice is that the method name used here is actually its fully qualified name (FQN). Again, this
has to do with being unambiguous. A method name, by itself, may refer to different methods, from different classes.
The FQN, on the other hand, is unique.

The third point is related to the arguments passed to this method. The signature of `fromAccessPointId` has three
parameters: `scope`, `id` and  `accessPointId`, but we are only passing the third one: the access point ID. The first
two are
inferred by deCDK; for the scope, deCDK uses a stack that it creates during the processing of the template to which
all constructs are added; for the ID, it uses the logical ID of the construct being declared (`"MyAccessPoint"` in
this case).

> If you need to provide these arguments explicitly, you can do it using two new (CDK specific) intrinsics:
> `CDK::Args` and `CDK::Scope`. Rewriting the example above, but with all arguments being explicitly
> defined, it becomes:

Reviewer: ```suggestion
If you need to provide these arguments explicitly, you can do it using the new (CDK specific) intrinsic function
`CDK::Args` and pseudo parameter `CDK::Scope`. Rewriting the example above, but with all arguments being explicitly
defined, it becomes:
```

is the method's simple name, rather than its FQN.

> ⚠️ **Experimental feature**: the string `On` is a keyword in YAML, equivalent to `true`. So you have to wrap it in
> quotes to use it as a property name. Since this is a bit cumbersome, expect this name to change in the near future.

You can also call instance method in _nested properties_:

```yaml
Resources:
  AllowFromEverywhere:
    'On': Listener # of type aws-cdk-lib.aws_elasticloadbalancingv2.ApplicationListener
    Call:
      connections.allowDefaultPortFromAnyIpv4:
        - Open to the world
```

### CDK classes that are not constructs

In a CDK application, many of the objects you create and manipulate are not constructs; they are not added
directly to a stack, and they don't correspond to a CloudFormation resource. For instance, to create an
`aws-cdk-lib.aws_cloudwatch.Dashboard`, which is a construct, the CDK construct library provides the interface
`aws-cdk-lib.aws_cloudwatch.IWidget`, with many non-construct implementations (`AlarmStatusWidget`, `TextWidget` etc).
These classes abstract
certain concepts and improve the ergonomics of the API.

> If the class has a public constructor with less than two required parameters (e.g., `TextWidget`), you can pass the
> argument via `Properties`:

Reviewer: ```suggestion
If the class has a public constructor with one required parameter that is a props interface (e.g., `TextWidget`), you can pass them via `Properties`:
```

```json
{
  "Resources": {
    "TitleWidget": {
      "Type": "aws-cdk-lib.aws_cloudwatch.TextWidget",
      "Properties": {
        "markdown": "# Operational Metrics"
      }
    },
    "ServiceDashboard": {
      "Type": "aws-cdk-lib.aws_cloudwatch.Dashboard",
      "Properties": {
        "widgets": [
          [
            {
              "Ref": "TitleWidget"
            }
          ]
        ]
      }
    }
  }
}
```

> > **Work in progress**: it is possible to support public constructors with two or more required parameters, but it
> > requires careful design of the syntax. Still under investigation.

Reviewer: ```suggestion
If no parameters are required, `Properties` can be omitted completely.

> **Work in progress**: it is possible to support public constructors with two or more required parameters, but it
> requires careful design of the syntax. Still under investigation.
```

Notice the use of the `Ref` intrinsic function, that works basically the same way as in raw CloudFormation resources,
although there are some nuances to this function in the context of deCDK, that are explained in **References**.

Non-construct objects can also be created with method calls:

```yaml
Resources:
  BusinessLogic:
    Call:
      aws-cdk-lib.aws_lambda.Code.fromInline: "exports.handler = function(event, ctx, cb) { return cb(null, \"hi\"); }"
```

> So far, we have seen how to create non-construct instances at the top level, with associated logical IDs, just like you

Reviewer: ```suggestion
#### Inline calls

So far, we have seen how to create non-construct instances at the top level, with associated logical IDs, just like you
```

```

### References

[//]: # (TODO Expand on this)

The intrinsic function `Ref` can be used to reference any logical ID (the keys in the `Resources` section), with two
exceptions:

* When the value of the logical ID is the result of a `void` method call. For example:

```yaml
Resources:
  ConfigureAsyncInvokeStatement:
    'On': Alias # of type aws-cdk-lib.aws_lambda.Alias
    Call:
      configureAsyncInvoke:
        retryAttempts: 2
```

In this case, `ConfigureAsyncInvokeStatement` doesn't refer to anything. It is only necessary to keep the syntax for
statements consistent with the other types of method calls.

* When a CloudFormation resource or an L1 construct is referenced where an L2 construct is expected:

```yaml
Resources:
  Bucket:
    Type: "AWS::S3::Bucket"

  MyFunction:
    Type: "aws-cdk-lib.aws_lambda.Function",
    Properties:
      code:
        Ref: "Bucket" # an aws_s3.Bucket is expected here
      # ... other properties
```

> **Work in progress**: We are investigating the possibility of supporting the second case.

Notably, CDK constructs, that is, types that extend `cdk.Construct` or interfaces that extend `cdk.IConstruct`, can be

> referenced from anywhere, even from CloudFormation resource declarations:

Reviewer: ```suggestion
referenced from anywhere, even from CloudFormation resource declarations.
In this case a reference to the underlying L1 CloudFormation resource is returned.
```
